velox/functions/sparksql/DecimalArithmetic.cpp

```diff
@@ -14,498 +14,360 @@
  * limitations under the License.
  */
 
-#include "velox/common/base/CheckedArithmetic.h"
 #include "velox/expression/DecodedArgs.h"
 #include "velox/expression/VectorFunction.h"
+#include "velox/functions/Macros.h"
+#include "velox/functions/Registerer.h"
 #include "velox/functions/sparksql/DecimalUtil.h"
 #include "velox/type/DecimalUtil.h"
 
 namespace facebook::velox::functions::sparksql {
 namespace {
 
-std::string getResultScale(std::string precision, std::string scale) {
-  return fmt::format(
-      "({}) <= 38 ? ({}) : max(({}) - ({}) + 38, min(({}), 6))",
-      precision,
-      scale,
-      scale,
-      precision,
-      scale);
-}
-
-// Returns the whole and fraction parts of a decimal value.
-template <typename T>
-inline std::pair<T, T> getWholeAndFraction(T value, uint8_t scale) {
-  const auto scaleFactor = velox::DecimalUtil::kPowersOfTen[scale];
-  const T whole = value / scaleFactor;
-  return {whole, value - whole * scaleFactor};
-}
-
-// Increases the scale of input value by 'delta'. Returns the input value if
-// delta is not positive.
-inline int128_t increaseScale(int128_t in, int16_t delta) {
-  // No need to consider overflow as 'delta == higher scale - input scale', so
-  // the scaled value will not exceed the maximum of long decimal.
-  return delta <= 0 ? in : in * velox::DecimalUtil::kPowersOfTen[delta];
+// Return precision of `aType` and `bType`, set the `aScale` and `bScale`.
+std::pair<int, int> getDecimalPrecisionScales(
```

Comment:
Perhaps use a new struct and set the member variables in `initialize` only when needed, seeing https://github.com/facebookincubator/velox/commit/cc48eaa5be8387e8988df8736b846454511e259f.
```
struct BinaryPrecisionScales {
  uint8_t aPrecision;
  uint8_t aScale;
  uint8_t bPrecision;
  uint8_t bScale;
};

// Extracts the precisions and scales for two decimal types.
BinaryPrecisionScales getDecimalPrecisionScales(
    const TypePtr& aType,
    const TypePtr& bType) {
  auto [aPrecision, aScale] = getDecimalPrecisionScale(*aType);
  auto [bPrecision, bScale] = getDecimalPrecisionScale(*bType);
  BinaryPrecisionScales result{
      static_cast<uint8_t>(aPrecision),
      static_cast<uint8_t>(aScale),
      static_cast<uint8_t>(bPrecision),
      static_cast<uint8_t>(bScale)};
  return result;
}
```